python/lsst/daf/butler/tests/butler_query.py

```diff
@@ -0,0 +1,1228 @@
+            LookupError, "Dimension element name cannot be inferred in this context."
+        ):
+            butler.query.data_ids(["detector"], where="timespan.end < time", bind=bind)
+
+    def testQueryDataIdsGovernorExceptions(self) -> None:
+        """Test exceptions raised by query.data_ids for incorrect governors."""
+        butler = self.make_butler("base.yaml", "datasets.yaml", "spatial.yaml")
+
+        def do_query(
+            dimensions: Iterable[str],
+            dataId: DataId | None = None,
+            where: str = "",
+            bind: dict | None = None,
+            **kwargs: Any,
+        ) -> DataCoordinateQueryResults:
+            return butler.query.data_ids(dimensions, data_id=dataId, where=where, bind=bind, **kwargs)
+
+        Test = namedtuple(
+            "Test",
+            ("dimensions", "dataId", "where", "bind", "kwargs", "exception", "count"),
+            defaults=(None, None, None, {}, None, 0),
+        )
+
+        test_data = (
+            Test("tract,visit", count=6),
+            Test("tract,visit", kwargs={"instrument": "Cam1", "skymap": "SkyMap1"}, count=6),
+            Test(
+                "tract,visit", kwargs={"instrument": "Cam2", "skymap": "SkyMap1"}, exception=DataIdValueError
+            ),
+            Test("tract,visit", dataId={"instrument": "Cam1", "skymap": "SkyMap1"}, count=6),
+            Test(
+                "tract,visit", dataId={"instrument": "Cam1", "skymap": "SkyMap2"}, exception=DataIdValueError
+            ),
+            Test("tract,visit", where="instrument='Cam1' AND skymap='SkyMap1'", count=6),
+            Test("tract,visit", where="instrument='Cam1' AND skymap='SkyMap5'", exception=DataIdValueError),
+            Test(
+                "tract,visit",
+                where="instrument=cam AND skymap=map",
+                bind={"cam": "Cam1", "map": "SkyMap1"},
+                count=6,
+            ),
+            Test(
+                "tract,visit",
+                where="instrument=cam AND skymap=map",
+                bind={"cam": "Cam", "map": "SkyMap"},
+                exception=DataIdValueError,
+            ),
+        )
+
+        for test in test_data:
+            dimensions = test.dimensions.split(",")
+            if test.exception:
+                with self.assertRaises(test.exception):
+                    do_query(dimensions, test.dataId, test.where, bind=test.bind, **test.kwargs).count()
+            else:
+                query = do_query(dimensions, test.dataId, test.where, bind=test.bind, **test.kwargs)
+                self.assertEqual(query.count(discard=True), test.count)
+
+            # and materialize
+            if test.exception:
+                with self.assertRaises(test.exception):
+                    query = do_query(dimensions, test.dataId, test.where, bind=test.bind, **test.kwargs)
+                    with query.materialize() as materialized:
+                        materialized.count(discard=True)
```

Comment:
Is code coverage telling us that it's actually entering the context manager that's failing here, not the intended test?  Might be worth reordering the context managers to make sure we're checking exactly the line we want.

Reply:
It looks like exception is actually happening in `do_query`, so context is not even entered. I'll drop the whole `with` clause.